src/de/mod.rs

```diff
@@ -419,11 +447,180 @@ impl<'de, 'a, R: BufRead> de::Deserializer<'de> for &'a mut Deserializer<R> {
     }
 }
 
+/// A trait that borrows an XML reader that borrows from the input. For a &[u8]
+/// input the events will borrow from that input, whereas with a BufRead input
+/// all events will be converted to 'static, allocating whenever necessary.
+pub trait BorrowingReader<'i> where Self: 'i {
+    /// Return an input-borrowing event.
+    fn next(&mut self) -> Result<Event<'i>, DeError>;
+
+    /// Skips until end element is found. Unlike `next()` it will not allocate
+    /// when it cannot satisfy the lifetime.
+    fn read_to_end(&mut self, name: &[u8]) -> Result<(), DeError>;
+
+    /// A copy of the reader's decoder used to decode strings.
+    fn decoder(&self) -> Decoder {
+        Decoder
+    }
+}
+
+struct IoReader<R: BufRead> {
+    reader: Reader<R>,
+    buf: Vec<u8>,
+}
+
+impl<'i, R: BufRead + 'i> BorrowingReader<'i> for IoReader<R> {
+    fn next(&mut self) -> Result<Event<'static>, DeError> {
```

Comment:
Both work, since `'static` can be used anywhere a different lifetime is required. I could change it if you want, but I like making it obvious why it is able to satisfy `'i` (i.e., it's static so it doesn't care about it).